pkg/gather/clusterconfig/0_gatherer.go

```diff
@@ -32,50 +75,92 @@ func New(gatherKubeConfig *rest.Config, gatherProtoKubeConfig *rest.Config, metr
-		GatherClusterOAuth(g),
-		GatherClusterIngress(g),
-		GatherClusterProxy(g),
-		GatherCertificateSigningRequests(g),
-		GatherCRD(g),
-		GatherHostSubnet(g),
-		GatherMachineSet(g),
-		GatherInstallPlans(g),
-		GatherServiceAccounts(g),
-		GatherMachineConfigPool(g),
-		GatherContainerRuntimeConfig(g),
-		GatherStatefulSets(g),
-	)
-}
+	var errors []string
+	var gatherReport []interface{}
+
+	if len(gatherList) == 0 {
+		errors = append(errors, "no gather functions are specified to run")
+	}
 
-func (g *Gatherer) setClusterVersion(version *configv1.ClusterVersion) {
-	g.lock.Lock()
-	defer g.lock.Unlock()
-	if g.lastVersion != nil && g.lastVersion.ResourceVersion == version.ResourceVersion {
-		return
+	if contains(gatherList, gatherAll) {
+		gatherList = make([]string, 0, len(gatherFunctions))
+		for k := range gatherFunctions {
+			gatherList = append(gatherList, k)
+		}
 	}
-	g.lastVersion = version.DeepCopy()
+
+	for _, gatherId := range gatherList {
+		gFn, ok := gatherFunctions[gatherId]
+		if !ok {
+			errors = append(errors, fmt.Sprintf("unknown gatherId in config: %s", gatherId))
+			continue
+		}
+		gatherName := runtime.FuncForPC(reflect.ValueOf(gFn).Pointer()).Name()
+		klog.V(5).Infof("Gathering %s", gatherName)
+
+		start := time.Now()
+		records, errs := gFn(g)
+		elapsed := time.Since(start).Truncate(time.Millisecond)
+
+		klog.V(4).Infof("Gather %s took %s to process %d records", gatherName, elapsed, len(records))
+		gatherReport = append(gatherReport, gatherStatusReport{gatherName, elapsed, len(records), errs})
+
+		for _, err := range errs {
+			errors = append(errors, err.Error())
+		}
+		for _, record := range records {
+			if err := recorder.Record(record); err != nil {
+				errors = append(errors, fmt.Sprintf("unable to record %s: %v", record.Name, err))
+				continue
+			}
+		}
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+	}
+
+	// Creates the gathering performance report
+	if err := recordGatherReport(recorder, gatherReport); err != nil {
+		errors = append(errors, fmt.Sprintf("unable to record io status reports: %v", err))
+	}
+
+	if len(errors) > 0 {
+		sort.Strings(errors)
+		errors = uniqueStrings(errors)
+		return fmt.Errorf("%s", strings.Join(errors, ", "))
+	}
+	return nil
 }
 
-// ClusterVersion returns Version for this cluster, which is set by running version during Gathering
-func (g *Gatherer) ClusterVersion() *configv1.ClusterVersion {
-	g.lock.Lock()
-	defer g.lock.Unlock()
-	return g.lastVersion
+func recordGatherReport(recorder record.Interface, report []interface{}) error {
+	r := record.Record{Name: "insights-operator/gathers", Item: record.JSONMarshaller{Object: report}}
+	return recorder.Record(r)
 }
+
+func uniqueStrings(arr []string) []string {
+	var last int
+	for i := 1; i < len(arr); i++ {
+		if arr[i] == arr[last] {
+			continue
+		}
+		last++
+		if last != i {
+			arr[last] = arr[i]
+		}
+	}
+	if last < len(arr) {
+		last++
+	}
+	return arr[:last]
+}
```

Comment:
Well I noticed that this is just copy/paste but does this function really work as expected? What is the expected return value for `uniqueStrings([]string{"b", "a", "a", "c", "a"}` ? I would expect `[b a c]` (or similar) but I am seeing `[b a c a]`. 

This https://github.com/openshift/insights-operator/blob/master/pkg/record/interface_test.go#L16 looks suspicious to me as well.

Reply:
yeah, this looks overly complicated for what it actually meant to do and that test case is just totally wrong (or its some moon logic that I don't understand)
should be easy to fix